Agent.skills dynamic callback ignores requestContextSchema — requestContext is always RequestContext<unknown>

## Description

When an agent defines `requestContextSchema`, every dynamic config callback (`instructions`, `tools`, `memory`, `model`, etc.) receives a typed `RequestContext<TRequestContext>` — except `skills`. The skills callback is always typed as `RequestContext<unknown>`, so key autocomplete and typo checking are lost.

Reported by a user on Discord as a follow-up to #19480 (that issue was about tool `execute` optionality, which is intentional; this is a separate, real typing gap).

## Reproduction (verified against `@mastra/core@1.51.0`)

```ts
import { Agent } from '@mastra/core/agent';
import { z } from 'zod';

new Agent({
  id: 'a',
  name: 'a',
  model: 'openai/gpt-4o-mini',
  requestContextSchema: z.object({ documentId: z.string(), userId: z.string() }),
  instructions: ({ requestContext }) => {
    requestContext.get('nonexistentKey'); // ❌ compile error — correctly typed
    return 'x';
  },
  skills: ({ requestContext }) => {
    requestContext.get('nonexistentKey'); // ✅ compiles silently — RequestContext<unknown>
    return [];
  },
});
```

`tsc --strict` output:

```
error TS2345: Argument of type '"nonexistentKey"' is not assignable to parameter of type '"documentId" | "userId"'.
```

— only for `instructions`. The identical call inside `skills` compiles.

## Root cause

`skills` is the only dynamic agent option that does not thread `TRequestContext`:

- `packages/core/src/agent/types.ts` — `skills?: AgentSkillsInput` (no generic), vs `memory?: DynamicArgument<MastraMemory, TRequestContext>` etc.
- `packages/core/src/skills/types.ts` — `AgentSkillsContext { requestContext: RequestContext }` falls back to the `unknown` default.

Runtime is unaffected — the real `RequestContext` is passed to the callback (`agent.ts`, `getSkillsWithWorkspace`). This is a compile-time-only gap.

## History / precedent

- #14581 / PR #14582 fixed this exact class of bug for 12 other dynamic fields (`model`, `memory`, `workflows`, `agents`, `scorers`, processors, ...) that were missed by the original `requestContextSchema` PR (#12259).
- PR #18360 later added agent-level skills with a bespoke `AgentSkillsContext` instead of `DynamicArgument`, reintroducing the same gap for the new field. Oversight, not a design decision.

## Proposed fix (type-only, backward compatible)

```ts
// packages/core/src/skills/types.ts
export interface AgentSkillsContext<TRequestContext extends Record<string, any> | unknown = unknown> {
  requestContext: RequestContext<TRequestContext>;
}

export type AgentSkillsInput<TRequestContext extends Record<string, any> | unknown = unknown> =
  | SkillInput[]
  | ((context: AgentSkillsContext<TRequestContext>) => SkillInput[] | Promise<SkillInput[]>);

// packages/core/src/agent/types.ts
skills?: AgentSkillsInput<TRequestContext>;
```

Plus the corresponding `#skills` private field in `agent.ts`. The `= unknown` defaults keep all existing usage compiling unchanged; no runtime change.

--- a/.changeset/shaggy-hornets-jam.md
+++ b/.changeset/shaggy-hornets-jam.md
@@ -1,0 +1,19 @@
+---
+'@mastra/core': patch
+---
+
+Fixed `requestContext` typing in the dynamic `skills` agent option. When an agent defines `requestContextSchema`, the `skills` callback now receives a typed `RequestContext` (with key autocomplete and typo checking), matching `instructions`, `tools`, `memory`, and the other dynamic options. Previously it was always `RequestContext<unknown>`.
+
+```typescript
+const agent = new Agent({
+  requestContextSchema: z.object({ documentId: z.string(), userId: z.string() }),
+  // Before: requestContext was RequestContext<unknown> — any key compiled
+  // After: requestContext is RequestContext<{ documentId: string; userId: string }>
+  skills: ({ requestContext }) => {
+    requestContext.get('documentId'); // typed as string
+    return ['./skills/basic'];
+  },
+});
+```
+
+Fixes [#19553](https://github.com/mastra-ai/mastra/issues/19553)

--- a/packages/core/src/agent/agent-types.test-d.ts
+++ b/packages/core/src/agent/agent-types.test-d.ts
@@ -151,6 +151,40 @@ describe('Agent Type Tests', () => {
 
       expectTypeOf(config.id).toEqualTypeOf<'test-agent'>();
     });
+
+    it('should type requestContext in skills function based on requestContextSchema', () => {
+      const config: AgentConfig<
+        'test-agent',
+        Record<string, never>,
+        undefined,
+        { documentId: string; userId: string }
+      > = {
+        id: 'test-agent',
+        name: 'Test Agent',
+        model: {} as any,
+        requestContextSchema: z.object({
+          documentId: z.string(),
+          userId: z.string(),
+        }),
+        instructions: 'You are a helpful assistant',
+        skills: ({ requestContext }) => {
+          // Verify requestContext is typed
+          expectTypeOf(requestContext).toEqualTypeOf<RequestContext<{ documentId: string; userId: string }>>();
+
+          // Verify get() returns the correct type
+          const documentId = requestContext.get('documentId');
+          expectTypeOf(documentId).toEqualTypeOf<string>();
+
+          // Verify unknown keys are rejected
+          // @ts-expect-error - key does not exist in the request context schema
+          requestContext.get('nonexistentKey');
+
+          return [];
+        },
+      };
+
+      expectTypeOf(config.id).toEqualTypeOf<'test-agent'>();
+    });
   });
 
   describe('Issue #16732: AgentExecutionOptions<undefined> should not require structuredOutput', () => {

--- a/packages/core/src/agent/agent-types.test-d.ts
+++ b/packages/core/src/agent/agent-types.test-d.ts
@@ -153,12 +153,9 @@ describe('Agent Type Tests', () => {
     });
 
     it('should type requestContext in skills function based on requestContextSchema', () => {
-      const config: AgentConfig<
-        'test-agent',
-        Record<string, never>,
-        undefined,
-        { documentId: string; userId: string }
-      > = {
+      // No explicit TRequestContext generic — the type must be inferred from
+      // requestContextSchema so a regression to RequestContext<unknown> fails.
+      new Agent({
         id: 'test-agent',
         name: 'Test Agent',
         model: {} as any,
@@ -181,9 +178,7 @@ describe('Agent Type Tests', () => {
 
           return [];
         },
-      };
-
-      expectTypeOf(config.id).toEqualTypeOf<'test-agent'>();
+      });
     });
   });
 

--- a/packages/core/src/agent/agent.ts
+++ b/packages/core/src/agent/agent.ts
@@ -487,7 +487,7 @@ export class Agent<
   #pubsub?: PubSub;
   #inheritedPubSub?: PubSub;
   #memory?: DynamicArgument<MastraMemory, TRequestContext>;
-  #skills?: AgentSkillsInput;
+  #skills?: AgentSkillsInput<TRequestContext>;
   #skillsFormat?: SkillFormat;
   #workflows?: DynamicArgument<Record<string, AnyWorkflow>, TRequestContext>;
   #defaultGenerateOptionsLegacy: DynamicArgument<AgentGenerateOptions, TRequestContext>;
@@ -1140,7 +1140,7 @@ export class Agent<
     if (this.#skills) {
       let resolvedInputs: SkillInput[];
       if (typeof this.#skills === 'function') {
-        resolvedInputs = await this.#skills({ requestContext: rc });
+        resolvedInputs = await this.#skills({ requestContext: rc as RequestContext<TRequestContext> });
       } else {
         resolvedInputs = this.#skills;
       }

--- a/packages/core/src/agent/types.ts
+++ b/packages/core/src/agent/types.ts
@@ -732,7 +732,7 @@ interface AgentConfigBase<
    * }
    * ```
    */
-  skills?: AgentSkillsInput;
+  skills?: AgentSkillsInput<TRequestContext>;
   /**
    * Format for skill information injection when workspace has skills.
    * @default 'xml'

--- a/packages/core/src/skills/types.ts
+++ b/packages/core/src/skills/types.ts
@@ -68,8 +68,8 @@ export type SkillInput = string | InlineSkill;
 /**
  * Context passed to a dynamic skills resolver.
  */
-export interface AgentSkillsContext {
-  requestContext: RequestContext;
+export interface AgentSkillsContext<TRequestContext extends Record<string, any> | unknown = unknown> {
+  requestContext: RequestContext<TRequestContext>;
 }
 
 /**
@@ -93,6 +93,8 @@ export interface AgentSkillsContext {
  * }
  * ```
  */
-export type AgentSkillsInput = SkillInput[] | ((context: AgentSkillsContext) => SkillInput[] | Promise<SkillInput[]>);
+export type AgentSkillsInput<TRequestContext extends Record<string, any> | unknown = unknown> =
+  | SkillInput[]
+  | ((context: AgentSkillsContext<TRequestContext>) => SkillInput[] | Promise<SkillInput[]>);
 
 export type { Skill, SkillMetadata, SkillFormat, WorkspaceSkills };

